kauri/tree.go

```diff
@@ -0,0 +1,170 @@
+package kauri
+
+import (
+	"math"
+
+	"github.com/relab/hotstuff"
+)
+
+// MaxChild determines the branching factor of the tree.
+const MaxChild = 2
+
+// TreeConfiguration is an abstraction for a tree communication model.
+type TreeConfiguration interface {
+	InitializeWithPIDs(ids map[hotstuff.ID]int)
+	GetHeight() int
+	GetChildren() []hotstuff.ID
+	GetSubTreeNodes() []hotstuff.ID
+	GetParent() (hotstuff.ID, bool)
+}
+
+// FaultFreeTree implements a fault free tree configuration.
+type FaultFreeTree struct {
+	ID                  hotstuff.ID
+	ConfigurationLength int
+	height              int
+	idToPosMapping      map[hotstuff.ID]int
+	posToIDMapping      map[int]hotstuff.ID
+}
+
+// CreateTree Creates the tree configuration, currently only fault free tree configuration is supported.
+func CreateTree(configurationLength int, myID hotstuff.ID) TreeConfiguration {
+
+	if configurationLength <= 0 {
+		return nil
+	}
+	temp := configurationLength
+	temp = temp - 1 //root
+	height := 1
+	for i := 1; temp > 0; i++ {
+		temp = temp - int(math.Pow(MaxChild, float64(i)))
+		height++
+	}
+	return &FaultFreeTree{
+		ID:                  myID,
+		ConfigurationLength: configurationLength,
+		height:              height,
+	}
+}
+
+// InitializeWithPIDs uses the map to initialize the position of replicas.
+func (t *FaultFreeTree) InitializeWithPIDs(ids map[hotstuff.ID]int) {
+	t.idToPosMapping = ids
+	t.posToIDMapping = make(map[int]hotstuff.ID)
+	for ID, index := range ids {
+		t.posToIDMapping[index] = ID
+	}
+}
+
+// GetParent fetches the ID of the parent, if root, returns itself.
+func (t *FaultFreeTree) GetParent() (hotstuff.ID, bool) {
+	myPos := t.idToPosMapping[t.ID]
+	if myPos == 0 {
+		return t.ID, false
+	}
+	return t.posToIDMapping[(myPos-1)/MaxChild], true
+}
+
+// GetChildren returns the children of the replicas, if any.
+func (t *FaultFreeTree) GetChildren() []hotstuff.ID {
+	return t.GetChildrenOfNode(t.ID)
+}
+
+func (t *FaultFreeTree) isWithInIndex(position int) bool {
```

Comment:
Is there a better name for this? Could this be simply `contains(index int) bool`?